### PR TITLE
Add atomic min and max

### LINKED ISSUE #1718: Implement atomic_min/max
In https://github.com/rust-lang/rust/pull/82391 I had to disable some tests because Miri does not support atomic_min/max.

In https://github.com/rust-lang/miri/pull/1653, an attempt was made to implement those intrinsics. The general approach looks good but see that PR for some of the remaining issues that need to be solved before landing this (also, there'll be lots of conflicts due to https://github.com/rust-lang/miri/pull/1716).

### PR DESCRIPTION
**Pull request by henryboisdequin**
Closes #1718
Previous attempt: #1653

TODO:

- [x] Merge `atomic_op` and `atomic_min_max` functions
- [x] Fix CI

**Note:** this PR also removes arbitrary trailing whitespace and generally formats the affected files

### CODE REVIEW

**Comment by RalfJung** [comment]
Thanks a lot for the PR. :)

> this PR also removes arbitrary trailing whitespace and generally formats the affected files

This causes quite a lot of noise, which makes review harder. Can you make the formatting changes a separate commit in the PR, or (even better) create a separate format-only PR? (Please don't rustfmt the entire repository, only these two files. There are some other files where rustfmt does a bad job, so I am waiting for some rustfmt fixes before formatting the entire repository.)

**Comment by henryboisdequin** [comment]
@RalfJung As you suggested in [this comment](https://github.com/rust-lang/miri/pull/1653#discussion_r545980341), I've merged the `atomic_op` and `atomic_min_max` functions by using this enum:

```rust
enum AtomicOp {
    MirOp(mir::BinOp, bool),
    Max,
    Min,
}
```

Also, how do I fix the CI errors (I'm not sure how to should I implement the trait the diagnostic suggested -the code with errors was from the previous PR-)?

File: src/data_race.rs [hunk 1]
```diff
@@ -542,6 +542,35 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: MiriEvalContextExt<'mir, 'tcx> {
         Ok(old)
     }
 
+    /// Perform an conditional atomic exchange with a memory place and a new
+    /// scalar value, the old value is returned.
+    fn atomic_min_max_scalar(
+        &mut self,
+        place: MPlaceTy<'tcx, Tag>,
```
Comment (oli-obk) [inline src/data_race.rs:549]: out of performance reasons, lots of code moved to using `&MplaceTy` and `&OpTy`, please also use this here

File: src/data_race.rs [hunk 2]
```diff
@@ -542,6 +542,35 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: MiriEvalContextExt<'mir, 'tcx> {
...
+        rhs: ImmTy<'tcx, Tag>,
+        min: bool,
+        atomic: AtomicRwOp,
+    ) -> InterpResult<'tcx, ImmTy<'tcx, Tag>> {
+        let this = self.eval_context_mut();
+
+        let old = this.allow_data_races_mut(|this| this.read_immediate(place.into()))?;
```
Comment (oli-obk) [inline src/data_race.rs:556]: This and the second call below must pass a `&OptTy`, so you'll need to take a reference of the result of the `.into()`

**Comment by RalfJung** [comment]
(I'm afraid I will probably not be able to take a look at the CI errors at least until the weekend, I am quite busy until next Tuesday.)

File: src/data_race.rs (hunk 1 above)
Comment (henryboisdequin) [reply src/data_race.rs]: @oli-obk So `place` would be a `&OpTy<'tcx, Tag>`

Comment (RalfJung) [reply src/data_race.rs]: `&MPlaceTy` makes more sense I think.

**Comment by henryboisdequin** [comment]
@oli-obk Do you know what the problem is?

Error from CI:

```
normalized stderr:
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `8192`,
 right: `4096`', $DIR/atomic.rs:88:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


expected stderr:


diff of stderr:

+thread 'main' panicked at 'assertion failed: `(left == right)`
+  left: `8192`,
+ right: `4096`', $DIR/atomic.rs:88:5
+note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
+

The actual stderr differed from the expected stderr.
Actual stderr saved to /tmp/compiletestswsV4Q/atomic.stderr
To update references, run this command from build directory:
tests/run-pass/update-references.sh '/tmp/compiletestswsV4Q' 'atomic.rs'

error: 1 errors occurred comparing output.
status: exit code: 101
command: "/home/runner/work/miri/miri/target/x86_64-unknown-linux-gnu/release/miri" "tests/run-pass/atomic.rs" "-L" "/tmp/compiletestswsV4Q" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-C" "prefer-dynamic" "-o" "/tmp/compiletestswsV4Q/atomic.stage-id" "-A" "unused" "--edition" "2018" "-Dwarnings" "-Dunused" "--sysroot" "/home/runner/.cache/miri/HOST" "-L" "/tmp/compiletestswsV4Q/atomic.stage-id.aux"
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `8192`,
 right: `4096`', tests/run-pass/atomic.rs:88:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

------------------------------------------

test [ui] run-pass/atomic.rs ... FAILED
```

File: tests/run-pass/atomic.rs [hunk 3]
```diff
@@ -74,6 +74,20 @@ fn atomic_u64() {
...
+    assert_eq!(ATOMIC.fetch_max(0x1000, SeqCst), 0x1000);
+    assert_eq!(ATOMIC.fetch_max(0x2000, SeqCst), 0x1000);
+    assert_eq!(ATOMIC.fetch_max(0x2000, SeqCst), 0x2000);
+
+    assert_eq!(ATOMIC.fetch_min(0x2000, SeqCst), 0x2000);
+    assert_eq!(ATOMIC.fetch_min(0x2000, SeqCst), 0x2000);
+    assert_eq!(ATOMIC.fetch_max(0x1000, SeqCst), 0x2000);
+    assert_eq!(ATOMIC.fetch_max(0x1000, SeqCst), 0x1000);
+    assert_eq!(ATOMIC.fetch_max(0x100, SeqCst), 0x1000);
+    assert_eq!(ATOMIC.fetch_max(0x10, SeqCst), 0x100);
```
Comment (oli-obk) [inline tests/run-pass/atomic.rs:90]:
```suggestion
    assert_eq!(ATOMIC.fetch_min(0x1000, SeqCst), 0x2000);
    assert_eq!(ATOMIC.fetch_min(0x1000, SeqCst), 0x1000);
    assert_eq!(ATOMIC.fetch_min(0x100, SeqCst), 0x1000);
    assert_eq!(ATOMIC.fetch_min(0x10, SeqCst), 0x100);
```

I think you just had a copy paste mistake, the function was actually doing the right thing

Comment (henryboisdequin) [reply tests/run-pass/atomic.rs]: Hmmm, the CI still fails

File: src/shims/intrinsics.rs [hunk 4]
```diff
@@ -295,110 +304,345 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
+            "atomic_umax" =>
+                this.atomic_op_min_max(args, dest, AtomicOp::Min, AtomicRwOp::SeqCst)?,
+            "atomic_umax_acq" =>
+                this.atomic_op_min_max(args, dest, AtomicOp::Min, AtomicRwOp::Acquire)?,
+            "atomic_umax_rel" =>
+                this.atomic_op_min_max(args, dest, AtomicOp::Min, AtomicRwOp::Release)?,
+            "atomic_umax_acqrel" =>
+                this.atomic_op_min_max(args, dest, AtomicOp::Min, AtomicRwOp::AcqRel)?,
+            "atomic_umax_relaxed" =>
+                this.atomic_op_min_max(args, dest, AtomicOp::Min, AtomicRwOp::Relaxed)?,
```
Comment (oli-obk) [inline src/shims/intrinsics.rs:616]: more copy paste swapped args

File: src/data_race.rs [hunk 5]
```diff
@@ -542,6 +542,35 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: MiriEvalContextExt<'mir, 'tcx> {
...
+        let this = self.eval_context_mut();
+
+        let old = this.allow_data_races_mut(|this| this.read_immediate(&place.into()))?;
+
+        // `binary_op` will bail if either of them is not a scalar.
+        let cond = if min {
+            this.overflowing_binary_op(mir::BinOp::Lt, &old, &rhs)?.0
+        } else {
+            this.overflowing_binary_op(mir::BinOp::Gt, &old, &rhs)?.0
+        };
```
Comment (oli-obk) [inline src/data_race.rs:563]: I think this is swapped. Afaict if it's a min operation, when `old < rhs` we should not overwrite the value, but the current code does it the other way around.

**Review by oli-obk — CHANGES_REQUESTED** [review]

File: src/data_race.rs (hunk 5 above)
Comment (henryboisdequin) [reply src/data_race.rs:563]: @oli-obk What should we do instead of overwriting the value when it's a min operation?

Comment (RalfJung) [reply src/data_race.rs:563]: I think the conclusion from the discussion in https://github.com/rust-lang/miri/pull/1653 was that this should *always* overwrite the value. Just like `fetch_add(0)` is always a write access (for the purpose of data race detection), so is `fetch_max(usize::MAX)`.

Comment (oli-obk) [reply src/data_race.rs:563]: well, you still need to find out the value of `cond` to know whether to use the new or the old value to write. In either case, the branches on this `if` condition are swapped. This does a `min` operation for `max` and vice versa.

Comment (henryboisdequin) [reply src/data_race.rs:563]: @oli-obk Do you mean that `!cond` then we `this.allow_data_races_mut(|this| this.write_immediate_to_mplace(*rhs, place))?`?

Comment (RalfJung) [reply src/data_race.rs:563]: I think the code should be more like
```rust
let lt = this.overflowing_binary_op(mir::BinOp::Lt, &old, &rhs)?.0.to_bool()?;
let new_val = if min {
  if lt { &old } else { &rhs }
} else /* max */ {
  if lt { &rhs } else { &old }
};
this.allow_data_races_mut(|this| this.write_immediate_to_mplace(new_val, place))?;
```

**Review by oli-obk — APPROVED** [review]

**Comment by henryboisdequin** [comment]
@oli-obk CI still fails though...

**Comment by oli-obk** [comment]
Yea, that's a regular build error, I approved the general code. Looks like you have references when a noon-reference was expected and an `ImmTy` instead of an `Immedidate`. If you deref an `ImmTy` you get an `Immediate`.

**Comment by henryboisdequin** [comment]
@oli-obk Understood! I didn't know that, thanks. I've changed the code accordingly.

**Comment by oli-obk** [comment]
Just out of curiosity: you are not building locally at all?

**Comment by henryboisdequin** [comment]
> Just out of curiosity: you are not building locally at all?

@oli-obk No not really, I will next time though. Not sure why I'm not.

**Comment by RalfJung** [comment]
> > this PR also removes arbitrary trailing whitespace and generally formats the affected files
> 
> This causes quite a lot of noise, which makes review harder. Can you make the formatting changes a separate commit in the PR, or (even better) create a separate format-only PR? (Please don't rustfmt the entire repository, only these two files. There are some other files where rustfmt does a bad job, so I am waiting for some rustfmt fixes before formatting the entire repository.)

This concern is still open, isn't it? Review is made difficult by having formatting changes and actual behavior changes in the same commit.

**Comment by oli-obk** [comment]
> Well, I kind of do have a reason: it takes more time to build locally than to look at the CI.

from scratch I understand that, but if you have that on every build, we should look into that.

Also, you still have some compilation failures.

**Comment by RalfJung** [comment]
@henryboisdequin We described in our contributors guide [how to set up an environment for local builds](https://github.com/rust-lang/miri/blob/master/CONTRIBUTING.md#preparing-the-build-environment). If you have any questions about that, feel free to ask. :)

**Comment by henryboisdequin** [comment]
> > > this PR also removes arbitrary trailing whitespace and generally formats the affected files
> > 
> > 
> > This causes quite a lot of noise, which makes review harder. Can you make the formatting changes a separate commit in the PR, or (even better) create a separate format-only PR? (Please don't rustfmt the entire repository, only these two files. There are some other files where rustfmt does a bad job, so I am waiting for some rustfmt fixes before formatting the entire repository.)
> 
> This concern is still open, isn't it? Review is made difficult by having formatting changes and actual behavior changes in the same commit.

@RalfJung Do you the best way to put all the formatting changes into one commit efficently?

**Comment by RalfJung** [comment]
You could try starting with master, doing the formatting, committing that, and then rebasing the remaining changes on top of that and hope that git conflict resolution works well. There's more advanced alternatives but they are harder to explain.^^

**Comment by henryboisdequin** [comment]
@RalfJung Done, this is ready for review

**Comment by oli-obk** [comment]
The build is still failing, maybe the rebase caused some match arms to get duplicated?

**Comment by henryboisdequin** [comment]
@oli-obk Thanks for the reminder, I was wondering why the diff was larger than before but didn't look into it 🤔

**Comment by henryboisdequin** [comment]
Hmm, tests still fail @RalfJung @oli-obk. Same problem as at first

File: tests/run-pass/atomic.rs (hunk 3 above)
Comment (oli-obk) [inline tests/run-pass/atomic.rs:90]: The failure is because your rebase reintroduced the copy paste mistake that was fixed previously.

Comment (henryboisdequin) [reply tests/run-pass/atomic.rs]: Woah, still fails

Comment (henryboisdequin) [reply tests/run-pass/atomic.rs]: Whoops, I got the error

Comment (henryboisdequin) [reply tests/run-pass/atomic.rs]: Nvm

File: src/data_race.rs [hunk 6]
```diff
@@ -687,15 +715,12 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: MiriEvalContextExt<'mir, 'tcx> {
         }
     }
 
-    fn reset_vector_clocks(
```
Comment (oli-obk) [inline src/data_race.rs:690]: all changes below this line in this file are rustfmt changes. Did rustfmt not pick these up in the first commit?

**Review by oli-obk — COMMENTED** [review]
I think you can squash all commits after the second one into the second one, as they are just going back and forth

File: src/data_race.rs (hunk 6 above)
Comment (henryboisdequin) [reply src/data_race.rs:690]: Hmm, that must have been the case

**Comment by henryboisdequin** [comment]
@oli-obk Commits squashed ✅

**Pull request merged**